### DISCUSSION TITLE
Docker Container Action is not able to run node script

### DISCUSSION THREAD

**Discussion by Sonichigo** [community/community#86691] [Discussions] [answered]
### Select Topic Area

General

### Body

I have been trying to setup a GitHub Action that runs a single node script (called index.js).
```
const core = require('@actions/core');
const github = require('@actions/github');

async function run() {
  try {
    const owner = core.getInput('owner', { required: true });
    const repo = core.getInput('repo', { required: true });
    const pr_number = core.getInput('pr_number', { required: true });
    const token = core.getInput('token', { required: true });

    const octokit = new github.getOctokit(token);

    const { data: changedFiles } = await octokit.rest.pulls.listFiles({
      owner,
      repo,
      pull_number: pr_number,
    });

    let diffData = {
      addition: 0,
      deletions: 0,
      changes: 0
    };

    diffData = changedFiles.reduce((acc, file) => {
      acc.additions += file.additions;
      acc.deletions += file.deletions;
      acc.changes += file.changes;
      return acc;
    }, diffData);

    await octokit.rest.issues.createComment({
      owner,
      repo,
      issue_number: pr_number,
      body: `
        Pull request #${pr_number} has be updated with: \n
        - ${diffData.changes} changes \n
        - ${diffData.additions} additions \n
        - ${diffData.deletions} deletions
      `
    });

    for (const file of changedFiles) {
      const fileExtention = file.filename.split('.').pop();
      let label = '';
      switch(fileExtention) {
        case 'md':
          label = 'markdown';
          break;
        case 'js':
          label = 'javascript';
          break;
        case 'yml':
          label = 'yaml';
          break;
        case 'yaml':
          label = 'yaml';
          break;
        default:
          label = 'noextension';
      }
      await octokit.rest.issues.addLabels({
        owner,
        repo,
        issue_number: pr_number,
        labels: [label]
      });
    }
  } catch (error) {
    core.setFailed(error.message);
  }
}

run();
```
When GitHub Actions then runs the container after building it, it automatically sets the working directory to /github/workspace, which is why I have replicated this in my Dockerfile ( WORKDIR /github/workspace/ ).

```
node:internal/modules/cjs/loader:1147
  throw err;
  ^

Error: Cannot find module '/github/workspace/index.js'
    at Module._resolveFilename (node:internal/modules/cjs/loader:1144:15)
    at Module._load (node:internal/modules/cjs/loader:9[8](https://github.com/Sonichigo/GitHub-Action-s/actions/runs/7500548232/job/20419453468#step:3:9)5:27)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:135:12)
    at node:internal/main/run_main_module:28:4[9](https://github.com/Sonichigo/GitHub-Action-s/actions/runs/7500548232/job/20419453468#step:3:10) {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}

Node.js v20.[11](https://github.com/Sonichigo/GitHub-Action-s/actions/runs/7500548232/job/20419453468#step:3:12).0
```
This suggests that the file does not exist. The debug output I added when building the container shows me otherwise though (root directory shortened for legibility)

This is my dockerfile
```dockerfile
FROM node:lts

COPY package*.json ./

RUN npm install

COPY index.js /index.js
RUN ["chmod", "+x", "/index.js"]

RUN ls
ENTRYPOINT ["node", "/index.js"]
```

**Comment by Naimulnobel** [comment]
The error message "Cannot find module '/github/workspace/index.js'" indicates that Node.js is unable to locate the `index.js` file in the `/github/workspace/` directory. This is because in your Dockerfile, you're copying `index.js` to the root directory (`/`) instead of the `/github/workspace/` directory.

Here's how you can fix this:

1. Update the `COPY` command in your Dockerfile to copy `index.js` to the `/github/workspace/` directory:

```dockerfile
COPY index.js /github/workspace/index.js
```

2. Update the `RUN` command that changes the permissions of `index.js`:

```dockerfile
RUN ["chmod", "+x", "/github/workspace/index.js"]
```

3. Update the `ENTRYPOINT` command to run `index.js` from the `/github/workspace/` directory:

```dockerfile
ENTRYPOINT ["node", "/github/workspace/index.js"]
```

So, your updated Dockerfile should look like this:

```dockerfile
FROM node:lts

COPY package*.json ./

RUN npm install

COPY index.js /github/workspace/index.js
RUN ["chmod", "+x", "/github/workspace/index.js"]

RUN ls
ENTRYPOINT ["node", "/github/workspace/index.js"]
```

This should resolve the issue and allow your GitHub Action to run the `index.js` script as expected.

**Reply by Sonichigo** [reply]
Unfortunately the issue still persists !! https://github.com/orgs/community/discussions/86691#discussioncomment-8129654

**Reply by Naimulnobel** [reply]
The error message "Cannot find module '/github/workspace/index.js'" is indicating that Node.js is unable to locate the `index.js` file in the `/github/workspace` directory. This is because in your Dockerfile, you're copying `index.js` to the root directory (`/`) instead of the `/github/workspace` directory.

However, from your `ls` command output, it seems like `index.js` is indeed in the root directory (`/`), not in `/github/workspace`.

Here's the relevant line from your Dockerfile:

```dockerfile
COPY index.js /index.js
```

And here's the command you're trying to run:

```dockerfile
ENTRYPOINT ["node", "/github/workspace/index.js"]
```

To fix this issue, you should change the `ENTRYPOINT` command to run the `index.js` file from the root directory, like this:

```dockerfile
ENTRYPOINT ["node", "/index.js"]
```

This should resolve the issue and allow your GitHub Action to run the `index.js` script as expected.

**Reply by Sonichigo** [reply]
I updated the dockerfile with the same information! if i do cd /github/workspace/ i can see the my index.js present there

**Reply by Naimulnobel** [reply]
Based on your requirements, here is a suggested Dockerfile:

```dockerfile
# Use an official Node.js runtime as a parent image
FROM node:lts

# Set the working directory in the container to /github/workspace
WORKDIR /github/workspace

# Copy package.json and package-lock.json (if available) to the working directory
COPY package*.json ./

# Install any needed packages specified in package.json
RUN npm install

# Copy the current directory contents into the container at /github/workspace
COPY . .

# If necessary, make your app available to the world outside this container
# EXPOSE 8080

# Define the command to run your app using CMD which defines your runtime
# Here we use "node", followed by the path to a JavaScript file
CMD ["node", "index.js"]
```

This Dockerfile sets up a Node.js environment, installs your app's dependencies, and runs your app. It assumes that your `index.js` and `package.json` files are in the same directory as your Dockerfile. If that's not the case, you'll need to adjust the `COPY` commands accordingly.

**Comment by Sonichigo** [comment, marked as answer]
The above error message was due to the contents for Dockerfile as well as the permission scope of GitHub actions.

select settings -> Actions -> Workflow permissions -> Read and write permissions .

the problem was due to copying in root path and relative path, so i setup the working directory and copied the source code their and running in index.js in the same workdir. this seemed to work

updated dockerfile:-
```
FROM node:22-alpine

WORKDIR /app

COPY package*.json /app/

RUN cd /app/ && npm install -y

COPY . /app/

RUN ls

RUN chmod +x /app/entrypoint.sh

ENTRYPOINT ["/app/entrypoint.sh", "/app/index.js"]
```

based on above i updated the entrypoint.sh as well:-

```
#!/bin/sh -l

echo "Running action"

## verfiying the path
echo "$PWD" 

## cd is necessary to move to the workdir otherwise we will get error module not found
cd /app && node index.js
```
